Add lastmod values to the sitemap index

joemcgill opened the pull request:
This implements a performant way of showing `lastmod` values for sitemap pages in the sitemap index.

### Issue Number
This will fix #60.

### Description
It's not performant to calculate `lastmod` values dynamically when the site index is rendered, so this approach calculates the `lastmod` values asynchronously. `lastmod` values for each sitemap page is stored in an option with a key of `core_sitemaps_lasmod_{object-type}_{object-subtype}_{page_number}` which are checked whenever the sitemap index is rendered. If no value exists, a single job is scheduled to fill in that value. All values are updated twice daily rather than updating them dynamically each time a post, taxonomy archive, or user is changed.

The main changes include:

- Moves responsibility for storing sitemap pages from the registry to each sitemap provider (7a64d11), which also improves general performance.
- Calculate `lastmod` values asynchronously and store as options (7ee9014, 6b9bab6).
- Update all `lastmod` values via a scheduled task run twice daily (56c3c16, 4da5f94, and other cleanup).

### Type of change
Please select the relevant options:

- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Enhancement (change which improves an existing feature. E.g., performance improvement, docs update, etc.)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Steps to test
- Visit the sitemap index for the first time to see that all lastmod values are empty.
- Run `wp cron event list` to see that events have been scheduled for calculating lastmod values.
- Wait for the events to run, or manually run `wp cron event run --all` to clear the scheduled events and see that `lastmod` values have been updated.

### Acceptance criteria
- [X] My code follows WordPress coding standards.
- [X] I have performed a self-review of my own code.
- [ ] If the changes are visual, I have cross browser / device tested.
- [X] I have commented my code, particularly in hard-to-understand areas.
- [X] My changes generate no new warnings.
- [X] I have added test instructions that prove my fix is effective or that my feature works.

Commit:
Move sitemap pages from the registry to sitemap providers.

This changes the way sitemaps are registered so that we now only store sitemap providers to the registry and then only caculate the number of pages when the index is rendered, by moving the responsibility for providing sitemap entry data from the registry to the sitemap provider base class.

This ensures queries required for determining the max number of pages for each sitemap sub-type are only run when the sitemap is being rendered, rather than when sitemaps are registered. It also allows for us to pass sitemap entry data directly into the index render rather than relying on pulling sitemap URL information out of the registry, separating concerns.

Commit:
Calculate sitemap lastmod times asynchronously

This checks for a lastmod value from the options table for each sitemap page when the index is rendered and schedules a WP_Cron task if no value is found.

Commit:
Fix CS issues

Commit:
Asynchronously update lastmod values.

This adds a scheduled event that fires runs twice daily to update all `lastmod` values for sitemap pages.

Commit:
Don't duplicate async jobs for lastmod

Commit:
Remove todo comment

joemcgill commented on inc/class-core-sitemaps-provider.php:
```diff
@@ -194,15 +216,136 @@ public function get_sitemaps() {
+	 * @param string $subtype The object subtype if applicable, e.g., post type, taxonomy type.
+	 * @param int    $page    The page number.
+	 */
+	public function calculate_sitemap_lastmod( $type, $subtype, $page ) {
```
I've opened #91 to address confusing naming of parameters and properties that I'd like to see addressed, but is outside of the scope of this issue.

Commit:
Remove delay when scheduling an async event.

Commit:
Remove unnecessary void return doc.

joemcgill commented on inc/class-core-sitemaps-provider.php:
```diff
@@ -194,15 +216,133 @@ public function get_sitemaps() {
+			$total = $this->max_num_pages( $name );
+
+			for ( $page = 1; $page <= $total; $page ++ ) {
+				wp_schedule_single_event( time(), 'core_sitemaps_calculate_lastmod', array( $this->slug, $name, $page ) );
```
We could check for an already scheduled event here as well.

svandragt replied:
See above about the 500s delay.

joemcgill commented on inc/class-core-sitemaps-provider.php:
```diff
@@ -42,6 +42,23 @@ class Core_Sitemaps_Provider {
+	/**
+	 * Set up relevant rewrite rules, actions, and filters.
+	 */
+	public function setup() {
```
I've moved the setup for all actions performed by providers from the main sitemap class and into the provider itself to keep responsibilities clear.

svandragt commented on inc/class-core-sitemaps-provider.php:
```diff
@@ -194,15 +196,63 @@ public function get_sitemaps() {
+	 * @param int    $page The page of the sitemap.
+	 * @return string The composed URL for a sitemap entry.
+	 */
+	public function get_sitemap_url( $name, $page ) {
```
nice encapsulation!

svandragt commented on inc/class-core-sitemaps-index.php:
```diff
@@ -70,8 +70,16 @@ public function render_sitemap() {
+
+			$sitemaps = array();
+
+			// Build up the list of sitemap pages.
```
Don't think the comment is required, the code is very clear.

svandragt commented on inc/class-core-sitemaps-provider.php:
```diff
@@ -245,12 +256,52 @@ public function get_sitemap_url( $name, $page ) {
+
+		$suffix = implode( '_', array_filter( array( $type, $subtype, (string) $page ) ) );
+
+		update_option( "core_sitemaps_lasmod_$suffix", $times[0] );
```
This looks like an intentional misspelling of lastmod, it might be worth commenting on that as it caught me by surprise thinking it was a type; or revising the name of the option.

joemcgill replied:
Good catch! this is _not_ intentional, I just misspelled in 🤦‍♂

svandragt commented on inc/class-core-sitemaps-provider.php:
```diff
@@ -245,12 +256,52 @@ public function get_sitemap_url( $name, $page ) {
+		// If blank, schedule a job.
+		if ( empty( $lastmod ) && ! wp_doing_cron() ) {
+			wp_schedule_single_event( time() + 500, 'core_sitemaps_calculate_lastmod', array( $this->slug, $name, $page ) );
+		}
```
This can result in a blank lastmod value rendering into the sitemap. Consider setting a fallback value to the current date at midnight in W3c datetime, after scheduling a job, so that there is always a valid timestamp.

joemcgill replied:
This is the intended behavior. Since `lastmod` values for sitemap pages is an expensive operation, we want to show no value when we don't have a previously calculated value, so we aren't blocking rendering of the page. When this is the case, we calculate this value asynchronously so it's available the next time the index is displayed.

svandragt replied:
Ah I missed that `lastmod` is an optional element, so it can be omitted safely. All good.

svandragt commented on inc/class-core-sitemaps-provider.php:
```diff
@@ -295,9 +296,12 @@ public function calculate_sitemap_lastmod( $type, $subtype, $page ) {
-		} );
+		usort(
+			$times,
+			function( $a, $b ) {
```
This closure can be declared as static to improve preformance.

joemcgill replied:
I've simplified this using `wp_list_sort()` in e9b24b4.

svandragt commented on inc/class-core-sitemaps-provider.php:
```diff
@@ -308,6 +316,31 @@ function( $a, $b ) {
+
+			$total = $this->max_num_pages( $name );
+
+			for ( $page = 1; $page <= $total; $page ++ ) {
```
FYI (outside the scope of this PR) There is protential here to reduce code duplication by passing in a callback into an abstracted function that calls the callback for each page, as this code is identical to `get_sitemap_entries()`.

joemcgill replied:
Yeah, I thought of that as well.

svandragt commented on inc/class-core-sitemaps-provider.php:
```diff
@@ -279,7 +279,12 @@ public function get_sitemap_lastmod( $name, $page ) {
+			$event_args = array( $this->slug, $name, $page );
+
+			// Don't schedule a duplicate job.
+			if ( ! wp_next_scheduled( 'core_sitemaps_calculate_lastmod', $event_args ) ) {
```
It's good to be extra safe, although if the codex (https://codex.wordpress.org/Function_Reference/wp_schedule_single_event) is to be believed it's currently not needed:

>Note that scheduling an event to occur before 10 minutes after an existing event of the same name will be ignored, unless you pass unique values for $args to each scheduled event. See wp_next_scheduled() for more information.
>
>This behavior is subject to change, as the original intention of the code was to prevent scheduling two identical events within ten minutes of each other, not preventing scheduling of identical events until ten minutes after the next scheduled occurrence.

joemcgill replied:
`wp_schedule_single_event()` indeed takes care of making sure duplicates are not scheduled within 10 minutes +/- one that is already scheduled. In this case, I'm trying to avoid scheduling a duplicate job _ever_. Perhaps unnecessary, but that's why I've added this check.

svandragt commented on inc/class-core-sitemaps-provider.php:
```diff
@@ -283,7 +283,7 @@ public function get_sitemap_lastmod( $name, $page ) {
 			// Don't schedule a duplicate job.
 			if ( ! wp_next_scheduled( 'core_sitemaps_calculate_lastmod', $event_args ) ) {
-				wp_schedule_single_event( time() + 500, 'core_sitemaps_calculate_lastmod', $event_args );
+				wp_schedule_single_event( time(), 'core_sitemaps_calculate_lastmod', $event_args );
```
The `+ 500` prevents running multiple events within the same 500s time period as an event is scheduled to run, by removing this the code is relying on the behaviour described in the codex that is subject to change.

joemcgill replied:
I don't think this is accurate. All that `+ 500` does is change the timestamp from "now" to 500 seconds later than "now". It won't have any effect on running multiple events beyond what `wp_schedule_single_event()` does internally to avoid duplicates within 10 minutes. I don't see a need to delay the event by 500 seconds, which is why I removed that value (which I initially added to give me a buffer to validated that all the events were actually scheduled).

svandragt replied:
Ok, happy with you having looked into it.

svandragt commented on inc/class-core-sitemaps-provider.php:
```diff
@@ -42,6 +42,23 @@ class Core_Sitemaps_Provider {
+		add_action( 'core_sitemaps_update_lastmod_' . $this->slug, array( $this, 'update_lastmod_values' ) );
+
+		if ( ! wp_next_scheduled( 'core_sitemaps_update_lastmod_' . $this->slug ) && ! wp_installing() ) {
+			wp_schedule_event( time(), 'twicedaily', 'core_sitemaps_update_lastmod_' . $this->slug );
```
Consider making the schedule `twicedaily` filterable, as depending on the use case spidering twice a day is not enough, eg news. (assuming the sitemap's last modified date influences whether the sitemap page should be explored by the crawler).

svandragt reviewed (changes requested):
Hey Joe, I've been a bit strict in this review because I thought it would be worth erring on the side of caution here; so feel free to consider my suggestions and choose to not implement them.

On the whole I believe the design is sound and will be performant on shared hosting sites with a small number of sitemaps using wp-cron, as well as larger sites with a task scheduler like cavalcade to perform async updates.

Excellent work.

Commit:
Remove unnecessary comment.

Commit:
Fix misspelling in lastmod option name.

Commit:
Make lastmod update recurrence filterable.

This adds a hook named `core_sitemaps_lastmod_recurrence` which allows the recurrence value for the scheduled job that updates lastmod values to be filtered. The hook also passes the object type from the provider so different recurrence values can be set based on object type.

Commit:
Simplify sorting of timestamps using wp_list_sort

Commit:
Abstract functionality calculating sitemap data.

This adds a method the the base provider class named `get_sitemap_type_data()` which can be used to get the name and max number of pages for each sitemap type, which is useful when rendering all sitemap entries as well as scheduling lastmod updates for all pages.

Commit:
Fix PHPCS issues

joemcgill commented:
@svandragt this is ready for another review when you have a chance. Thanks for all the feedback!

svandragt commented:
This is such a nice project and the PR is a really good improvement 👍 Nice job @joemcgill
